### PR TITLE
Add a script to generate a new manifest

### PR DESCRIPTION
**Pull request by PengTian0**
**Background**
Jenkins will try to build debian packages, vagrant box, ova, docker every night.
All the packages should use the same source code of rackhd.
In order to make all the packages building with the same source code of rackhd, we create a job "CreateManifest" to generate a manifest file and upload it to bintray.
The jobs which build packages will download the manifest file from bintray ,  check out repositories according to the manifest, and build packages.

**Implementation**
1. update the  manifest sample with parameter "branch"
2. checkout all the on-xxx repositories according to the updated manifest.
3. get the last commit of each repository before yesterday 23:59:59 
4. update the manifest with the commit.
5. write the manifest to a file which naming with "branch name"-"date of yesterday" 

**Test Job**
http://rackhdci.lss.emc.com/job/Maglev-BRI-Test/job/CreateManifest/

### CODE REVIEW

**Comment by PengTian0** [comment]
@panpan0000 @changev @yyscamper @anhou

File: manifest-build-tools/application/manifest_generator.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+--git-credential https://github.com,GITHUB
+
+The required parameters: 
+source-manifest: The path of manifest. The new manifest is generated according to it.
+dest-manifest: The path of the new manifest
+branch: The new branch name
+
+The optional parameters:
+force: If true, overwrite the destination manifest file even it already exists.
+publish: If true, the script will try to push the new manifest to github. That means the dest manifest should         under the manifest repository.
```
Comment (panpan0000) [inline manifest-build-tools/application/manifest_generator.py]: extra spaces.. and what does it mean "dest manifest should under the manifest repo "?
if the dest-manifest is a new file name, will you create a new file ?

File: manifest-build-tools/application/manifest_generator.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+--source-manifest build-manifest/rackhd-devel \
+--dest-manifest build-manifest/rackhd-release-1.2.5 \
+--branch branch/release-1.2.5 \
+--force \
+--publish \
+--publish-branch master \
+--git-credential https://github.com,GITHUB
+
+The required parameters: 
+source-manifest: The path of manifest. The new manifest is generated according to it.
```
Comment (panpan0000) [inline manifest-build-tools/application/manifest_generator.py]: what is the source-file used for ?
in my mind, you can obtain all information  you need from git repos , no need the old manifest file ?

File: manifest-build-tools/application/manifest_generator.py (hunk 1 above)
Comment (PengTian0) [inline manifest-build-tools/application/manifest_generator.py]: It should be changed to "run git push under the directory of dest manifest".
if the dest-manifest is a new file name, I will create a new file.

File: manifest-build-tools/application/manifest_generator.py (hunk 2 above)
Comment (PengTian0) [inline manifest-build-tools/application/manifest_generator.py]: I need to add repositories to new manifest file according to the source-file.

File: manifest-build-tools/application/manifest_generator.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+dest-manifest: The path of the new manifest
+branch: The new branch name
+git-credential: Git credentials for CI services.
+builddir: The directory for checked repositories.
+
+The optional parameters:
+force: If true, overwrite the destination manifest file even it already exists.
+publish: If true, the script will try to run "git push" under the directory of dest manifest. 
+publish-branch: The new manifest will be pushed to the branch.
+jobs: number of parallel jobs to run. The number is related to the compute architecture, multi-core processors...
```
Comment (panpan0000) [inline manifest-build-tools/application/manifest_generator.py]: +jobs: number of parallel jobs to run( **when git-clone code**)....

**Comment by panpan0000** [comment]
:+1: 

If I were you , I may write description:


- Background
when creating a new branch for RackHD( typically will create a new release-branch for incoming release, for each on-xxx repo), several works need to be done:

  1. create branch
  2. update version in debian/changelog( typically for a new release, the version will bump)
  3. update package.json to reflect the on-core/on-tasks dependence from master to this new branch
  4. generate a new manifest for this new branch

this PR address above step #4.

- Summary of implementation
 1.  before this script invoked , all on-xxx repo should be cloned into "```build-dir```"
 2.  new branch should has been created.
 3.  based on a old manifest file (```source-manifest```) ,  pre-knowledge is obtained : example:  manifest layout format and which repos(their URL) will be processed
4.  new manifest will be generated : for each repo, the ```commit-id``` will be their latest tip, and ```branch``` will be the new branch. the repository URL will be the same as ```source-manifest```
5.  if ```publish```, this new manifest will be uploaded to target github repo

- Test Result:
blablablabla

File: manifest-build-tools/application/manifest_generator.py (hunk 1 above)
Comment (PengTian0) [inline manifest-build-tools/application/manifest_generator.py]: done！

**Comment by changev** [comment]
:+1:

**Comment by PengTian0** [comment]
@panpan0000 @changev

**Comment by panpan0000** [comment]
one more question, the https://dl.bintray.com/pengtia​n0/binary/master-xxxx  is available for download after several seconds after Jenkins job completed successfully.
I'm worried this time lag will bring problem running with its downstream job.

File: manifest-build-tools/application/generate_manifest.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+--force \
+--git-credential https://github.com,GITHUB \
+--jobs 8
+
+The required parameters: 
+branch: The branch name of each repository in manifest file.
+date: The commit of each repository are last commit before the date.
+      The valid date: current (the commit of each repository are the lastest commit")
+                      yesterday (the commit of each repository are the last commit of yesterday")
+                      date string, such as: 2016-12-01 00:00:00 (the commit of each repository are the last commit before the date")
```
Comment (PengTian0) [inline manifest-build-tools/application/generate_manifest.py]: I submitted a commit to add parameter timezone!

**Comment by PengTian0** [comment]
@panpan0000 
I downloaded the file successfully after uploading it to bintray.

**Comment by PengTian0** [comment]
@panpan0000 
Yes, this situation may happen.
I added code to try to download the manifest file 100 times after upload it to bintray.
If failed, the job will fail and won't trigger downstream jobs.

File: manifest-build-tools/lib/manifest.py [hunk 5]
```diff
@@ -12,7 +12,7 @@
 
 from gitbits import GitBit
 
-
+manifest_sample = "manifest.json"
```
Comment (panpan0000) [inline manifest-build-tools/lib/manifest.py]: still perfer it with better naming :-)   like manifest.json.sample or manifest.json.template

File: manifest-build-tools/lib/RepositoryOperator.py [hunk 6]
```diff
@@ -317,6 +317,16 @@ def get_lastest_commit_id(self, repo_dir):
         else:
             raise RuntimeError("Unable to get commit id in directory {0}".format(repo_dir))
 
+    def get_lastest_merge_commit_before_date(self, repo_dir, date):
+        if repo_dir is None or not os.path.isdir(repo_dir):
+            raise RuntimeError("The repository directory is not a directory")
+        return_code, output, error = self.git.run(['log', '--merges', '--format=format:%H', '--before='+date, '-n', '1'], directory=repo_dir)
```
Comment (panpan0000) [inline manifest-build-tools/lib/RepositoryOperator.py]: Thanks. I was worrying the ```git log --before="datestring"```,  there should be quote around $datestring.
but Tian told me the self.git.run() will add quote(") around those element . so it's fine.

**Comment by PengTian0** [comment]
@changev

**Pull request merged**